Add `set_config` method to RP SPI driver

dstric-aqueduct opened the pull request:
Add a `set_config` method to `Spi` struct to allow reconfiguring SPI mode after creation.

The existing implementation of the `embassy-embedded-hal` trait `SetConfig` is changed to use the new method.

Existing uses of `SetConfig`  trait may need to explicitly call the trait method to maintain current return type.

Dirbaio commented on embassy-rp/src/spi.rs:
```diff
@@ -244,6 +244,27 @@ impl<'d, T: Instance, M: Mode> Spi<'d, T, M> {
+
+        // change stuff
+        p.cpsr().write(|w| w.set_cpsdvsr(presc));
+        p.cr0().write(|w| {
```
this is copypasted from `new_inner`. could you extract it to a private function `fn apply_config()` instead?

dstric-aqueduct replied:
After the CI failure, I see that the `SetConfig` trait in the [embassy-embedded-hal](https://docs.embassy.dev/embassy-embedded-hal/git/default/trait.SetConfig.html) actually does what I want, so this change may not be needed. Should we consider changing the trait implementation to disable/reenable the peripheral as the `set_frequency` method does?

Dirbaio replied:
whatever's needed for it to work. I don't know if it's required

dstric-aqueduct replied:
The [Pico SDK disables the driver before changing frequencies](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_spi/spi.c), as does the `set_frequency` method. I've updated to add a private `apply_config` method and then reused the `Spi` struct's `set_config` method in the trait implementation.

dstric-aqueduct replied:
This would also be a breaking change as the method's could not conflict depending on Trait in scope or not - I propose we just remove the proposed Spi `set_config` method and update the Trait implementation to disable/reenable the driver at the expense of some code duplication - I'll wait for your thoughts before touching again. Thanks for all your work.

Dirbaio replied:
don't worry about breaking changes, nextreleasewill be breaking anyway due to other changes.

Dirbaio reviewed (approved):
thanks!

Dirbaio commented:
- Can you fix CI?
- Can you squash all commits, and give it a good description of what you changed? "Update spi.rs" is not descriptive at all. (I recommend you do PRs by cloning the repo and using a real desktop git client, the github web ui encourages terrible practices...)

dstric-aqueduct commented:
> * Can you fix CI?
> * Can you squash all commits, and give it a good description of what you changed? "Update spi.rs" is not descriptive at all. (I recommend you do PRs by cloning the repo and using a real desktop git client, the github web ui encourages terrible practices...)

Squashed all the commits and cleaned up the PR description. 

I think the CI is running cached tests so the modified `spi_sdmmc` example in the PR is not propagating to CI? Not sure how to sync this.

Dirbaio commented:
fixed+rebased it for you at https://github.com/embassy-rs/embassy/pull/3600

(next time please tick the "allow maintainers to push" box, so I can rebase without creating a new PR. Thanks!)
